feat(snap) add snap features

tonyespy opened the pull request:
This PR adds the following snap features:

- it updates the snap to use a build-snap for go (v.s. downloading a binary release)
- it adds snap content interfaces for vault tokens
- it adds environment var configuration override support (aka rich config hooks) 
- it uses git tags to determine the version
- it bumps to the epoch to 4, to prevent refreshes from a 1.x snap

## PR Checklist
Please check if your PR fulfills the following requirements:

- [NA] Tests for the changes have been added (for bug fixes / features)
- [NA] Docs have been added / updated (for bug fixes / features)

<!-- If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/.github/CONTRIBUTING.md -->

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->

Issue Number:


## What is the new behavior?


## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Are there any new imports or modules? If so, what are they used for and why?


## Are there any specific instructions or things that should be known prior to reviewing?

## Other information

siggiskulason commented:
Looks good so far - we just need to sort out the question of getting the vault secret-token and decide on a way forward with that, as discussed earlier.

lenny-intel commented on snap/README.md:
```diff
@@ -110,3 +130,41 @@ $ sudo sed -i -e 's@/vault/config/assets/resp-init.json@/var/snap/edgex-app-serv
+service.startup-msg             // Service.StartupMsg
+service.request-timeout         // Service.RequestTimeout
+
+[Clients.core-data]
+clients.core-data.host                 // Clients.core-data.Host
+clients.core-data.port                 // Clients.core-data.Port
+clients.core-data.protocol             // Clients.core-data.Protocol
```
Note with V2 we now have the possibility of additional Metadata, Command and Notification clients. Not require unless using the related Client interfaces. Also, Metadata has become the require Client for version check to work since Core Data is now optional.
I will be updating ASC profiles to have to appropriate clients.
Here is example from the template config TOML: https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/app-service-template/res/configuration.toml#L73-L92

lenny-intel commented on snap/local/hooks/Makefile:
```diff
@@ -0,0 +1,23 @@
+cmd/install/install:
+	$(GO) build -o $@ ./cmd/install
+
+# TODO: fixup unit testing
+test:
+	$(GO) test -coverprofile=coverage.out ./...
+	$(GO) vet ./...
+	gofmt -l .
+	[ "`gofmt -l .`" = "" ]
+#	./bin/test-go-mod-tidy.sh
+#	./bin/test-attribution-txt.sh
```
Do you need this since the main makefile already forced it to run and pass in the CI build?

tonyespy commented:
@lenny-intel thanks for the review. I updated the `snap/README.md` file re: the clients, and also cleaned up the commented out code in the test target of the `Makefile`. It didn't make sense to completely remove the target. Maybe we can chat about this on slack if you think it's not right?

lenny-intel commented:
@tonyespy , I am fine with leaving it,  but does any thing actual use it?

lenny-intel reviewed (dismissed):
LGTM

tonyespy commented:
PR revised and ready for review

lenny-intel reviewed (approved):
LGTM
